graph/include/LoggerUtil.h

```diff
@@ -32,11 +34,245 @@ class MCGLogger {
+   */
+  template <typename MSG_t, typename... Args>
+  static void logInfoUnique(const MSG_t msg, Args&&... args) {
+    metacg::MCGLogger::instance().info<LogType::UNIQUE, Output::StdConsole>(msg, std::forward<Args>(args)...);
+  }
+
+  /**
+   * A shorthand to print a error to the console
+   *
+   * @tparam MSG_t inferred from msg
+   * @tparam Args  inferred from args
+   * @param msg the msg to be formatted and printed
+   * @param args the arguments to format the message with
+   */
+  template <typename MSG_t, typename... Args>
+  static void logError(const MSG_t msg, Args&&... args) {
+    metacg::MCGLogger::instance().error<LogType::DEFAULT, Output::ErrConsole>(msg, std::forward<Args>(args)...);
+  }
+
+  /**
+   * A shorthand to print a unique error to the console
+   *
+   * @tparam MSG_t inferred from msg
+   * @tparam Args  inferred from args
+   * @param msg the msg to be formatted and printed
+   * @param args the arguments to format the message with
+   */
+  template <typename MSG_t, typename... Args>
+  static void logErrorUnique(const MSG_t msg, Args&&... args) {
+    metacg::MCGLogger::instance().error<LogType::UNIQUE, Output::ErrConsole>(msg, std::forward<Args>(args)...);
+  }
+
+  /**
+   * A shorthand to print a unique error to the console
+   *
+   * @tparam MSG_t inferred from msg
+   * @tparam Args  inferred from args
+   * @param msg the msg to be formatted and printed
+   * @param args the arguments to format the message with
+   */
+  template <typename MSG_t, typename... Args>
+  static void logWarn(const MSG_t msg, Args&&... args) {
+    metacg::MCGLogger::instance().warn<LogType::DEFAULT, Output::StdConsole>(msg, std::forward<Args>(args)...);
+  }
+
+  /**
+   * A shorthand to print a unique error to the console
+   *
+   * @tparam MSG_t inferred from msg
+   * @tparam Args  inferred from args
+   * @param msg the msg to be formatted and printed
+   * @param args the arguments to format the message with
+   */
+  template <typename MSG_t, typename... Args>
+  static void logWarnUnique(const MSG_t msg, Args&&... args) {
+    metacg::MCGLogger::instance().warn<LogType::UNIQUE, Output::StdConsole>(msg, std::forward<Args>(args)...);
+  }
+
  private:
   MCGLogger() : console(spdlog::stdout_color_mt("console")), errconsole(spdlog::stderr_color_mt("errconsole")) {}
 
+  template <LogType lt>
+  inline bool ensureUnique(const std::string& formattedMessage) {
+    if constexpr (lt == LogType::UNIQUE) {
+      const size_t msg_hash = std::hash<std::string>()(formattedMessage);
+      // If we found the msg, we just return
+      if (allreadyPrintedMessages.find(msg_hash) != allreadyPrintedMessages.end()) {
+        return true;
+      }
+      allreadyPrintedMessages.insert(std::hash<std::string>()(formattedMessage));
+    }
+    return false;
+  }
+
   std::shared_ptr<spdlog::logger> console;
   std::shared_ptr<spdlog::logger> errconsole;
+
+  std::unordered_set<size_t> allreadyPrintedMessages;
```

Comment:
```suggestion
  std::unordered_set<size_t> alreadyPrintedMessages;
```
Just found that typo during my last pass.